Doc/reference/datamodel.rst

```diff
@@ -2859,33 +2859,83 @@ through the object's keys; for sequences, it should iterate through the values.
+      a[(slice(1, 2, None),)]
+      a[(slice(1, 2, None), slice(2, 3, None))]
 
-   and so forth.  Missing slice items are always filled in with ``None``.
+   Missing slice items are always filled in with ``None``.
 
 
 .. method:: object.__getitem__(self, key)
 
-   Called to implement evaluation of ``self[key]``. For :term:`sequence` types,
-   the accepted keys should be integers. Optionally, they may support
-   :class:`slice` objects as well.  Negative index support is also optional.
-   If *key* is
-   of an inappropriate type, :exc:`TypeError` may be raised; if *key* is a value
-   outside the set of indexes for the sequence (after any special
-   interpretation of negative values), :exc:`IndexError` should be raised. For
-   :term:`mapping` types, if *key* is missing (not in the container),
-   :exc:`KeyError` should be raised.
+   Called to implement evaluation of ``self[key]``. There are 3 supported
+   duck types: :term:`sequence` types, :term:`mapping` types, and any type
+   with custom behaviour on indexing. Syntax doesn't perform any checks,
+   so it is this method responsibility to raise :exc:`TypeError` on keys of
+   improper type.
+
+   First, there is expected behaviour for :term:`sequence` types:.
+   Minimal requirement for :func:`iter` to work (without existing
+   :meth:`__iter__` method) is that, ``__getitem__`` needs to accept
+   integers from 0 up to the index, on which :exc:`IndexError` is raised,
+   stoping the iteration. :func:`reversed` will also call ``len(self)`` and
+   request only items at non-negative keys. However, if class passes as
+   :class:`collections.abc.Sequence`, ``__getitem__`` should also:.
+
+   1. Accept negative integer keys, for indexing from the end of the sequence.
+   2. For all integer keys outside of bounds ``-len(self) <= key < len(self)``,
+      raise :exc:`TypeError`.
+   3. For non-integer key that supports :func:`operator.index`, use that
+      index as integer key.
+   4. Fully support slicing. If possible, it should return the same type,
+      if not, it should return a similar type. Meaning, ``self[::]`` should
+      create shallow copy.
+
+   Second, there is expected behaviour for :term:`mapping` types:.
+   ``__getitem__`` should return element for given *key*, if *key* is missing
+   (not in the container), :exc:`KeyError` should be raised.
+   To avoid silent errors, ``__getitem__`` should only implement mapping
+   (or not implement it at all), but if needed, it could be extended
+   for *keys* of different type. For implicit (ABC) children of
+   :class:`collections.abc.Mapping`, static type checkers are allowed to
+   deduce *keys* type as all types accepted by ``__getitem__``.
+   To avoid such errors, :term:`mapping` types with extended ``__getitem__``
+   need to override this behaviour by explicitly deriving from
+   :ref:`types-genericalias` of :class:`collections.abc.Mapping`.
+
+   Lastly, here is an example of possible implementation for ``Matrix``,
```

Comment:
When you say "first, second, lastly", it appears like you are describing requirements but all of them are actually unrelated to each other. You can just start your paragraphs normally like "For Sequence types, ...", "For mapping types,..." and use an Example section.